auth, etcdserver: introduce revision of authStore for avoiding TOCTOU problem

mitake opened the pull request:
This commit introduces revision of authStore. The revision number
represents a version of authStore that is incremented by updating auth
related information.

The revision is required for avoiding TOCTOU problems. Currently there
are two types of the TOCTOU problems in v3 auth.

The first one is in ordinal linearizable requests with a sequence like
below ():
1. Request from client CA is processed in follower FA. FA looks up the
   username (let it U) for the request from a token of the request. At
   this time, the request is authorized correctly.
2. Another request from client CB is processed in follower FB. CB
   is for changing U's password.
3. FB forwards the request from CB to the leader before FA. Now U's
   password is updated and the request from CA should be rejected.
4. However, the request from CA is processed by the leader because
   authentication is already done in FA.

For avoiding the above sequence, this commit lets
etcdserverpb.RequestHeader have a member revision. The member is
initialized during authentication by followers and checked in a
leader. If the revision in RequestHeader is lower than the leader's
authStore revision, it means a sequence like above happened. In such a
case, the state machine returns auth.ErrAuthRevisionObsolete. The
error code lets nodes retry their requests.

The second one, a case of serializable range and txn, is more
subtle. Because these requests are processed in follower directly. The
TOCTOU problem can be caused by a sequence like below:
1. Serializable request from client CA is processed in follower FA. At
   first, FA looks up the username (let it U) and its permission
   before actual access to KV.
2. Another request from client CB is processed in follower FB and
   forwarded to the leader. The cluster including FA now commits a log
   entry of the request from CB. Assume the request changed the
   permission or password of U.
3. Now the serializable request from CA is accessing to KV. Even if
   the access is allowed at the point of 1, now it can be invalid
   because of the change introduced in 2.

For avoiding the above sequence, this commit lets the functions of
serializable requests (EtcdServer.Range() and EtcdServer.Txn())
compare the revision in the request header with the latest revision of
authStore after the actual access. If the saved revision is lower than
the latest one, it means the permission can be changed. Although it
would introduce false positives (e.g. changing other user's password),
it prevents the TOCTOU problem. This idea is an implementation of
Anthony's comment:
https://github.com/coreos/etcd/pull/5739#issuecomment-228128254

/cc @heyitsanthony @xiang90

heyitsanthony commented on auth/store.go:
```diff
@@ -110,23 +116,27 @@ type AuthStore interface {
-	UsernameFromToken(token string) (string, bool)
+	// AuthInfoFromToken gets a username from the given Token and current revision number
+	// (The revision number is used for preventing the TOCTOU problem)
+	AuthInfoFromToken(token string) (string, uint64, bool)
```
the user and rev are used together everywhere, it would probably be cleaner to have `type AuthUser struct { user string; rev int64}` instead of having them separate

mitake replied:
OK

mitake replied:
I'd like to use `AuthInfo` instead of `User` because `authpb` already has `User`. How do you think about the naming?

heyitsanthony commented on etcdserver/api/v3rpc/rpctypes/error.go:
```diff
@@ -113,6 +115,7 @@ var (
 	ErrPermissionDenied     = Error(ErrGRPCPermissionDenied)
 	ErrRoleNotGranted       = Error(ErrGRPCRoleNotGranted)
 	ErrPermissionNotGranted = Error(ErrGRPCPermissionNotGranted)
+	ErrRetryAuthRequired    = Error(ErrGRPCRetryAuthRequired)
```
I don't think this should be something ever returned by the server; if a retry is necessary the server can retry internally

mitake replied:
OK, I was thinking that where the retrying mechanism should be placed e.g. clientv3. But etcdserver would be a suitable place for the retrying.

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -81,15 +82,20 @@ func (s *EtcdServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeRe
+				Header: &pb.RequestHeader{Username: user, Revision: rev},
 				Range:  r})
+
+		if rev < s.authStore.Revision() {
```
this can just loop until `rev == s.authStore.Revision` instead of bailing out if there was an auth update

mitake replied:
OK

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -129,14 +135,21 @@ func (s *EtcdServer) Txn(ctx context.Context, r *pb.TxnRequest) (*pb.TxnResponse
+				Header: &pb.RequestHeader{Username: user, Revision: rev},
 				Txn:    r})
+
+		if rev < s.authStore.Revision() {
```
same here

mitake replied:
OK

heyitsanthony commented on auth/store.go:
```diff
@@ -765,3 +827,22 @@ func hasRootRole(u *authpb.User) bool {
 	return false
 }
+
+func putRevision(tx backend.BatchTx, rev uint64) {
```
putRevision is always paired with a revision bump, maybe instead:
```
func (as *authStore) commitRevision(tx backend.BatchTx) {
    as.revision++
    revBytes := make([]byte, revBytesLen)
    binary.BigEndian.PutUint64(revBytes, rev)
    tx.UnsafePut(authBucketName, revisionKey, revBytes)
}
```

mitake replied:
Seems nice, I'll follow your idea.

mitake commented:
@heyitsanthony updated for your comments, PTAL

mitake commented:
@heyitsanthony sorry for long stopping. I updated this PR and the CI failures would be resolved. I changed below points:
- invalidate tokens during user deletion and password update
- renamed the error code from `auth.ErrRetryAuthRequired` to `auth.ErrAuthRevisionObsolete`. Based on the error code, linearizable requests retry until they see success or other error cases (currently some requests e.g. lease related are not check the error code because auth cannot care about them for now)
- serializable requests check the revision numbers in their headers after processing
- removed a bug that causes nil pointer dereference by serialized request when auth isn't enabled

Could you take a look?

heyitsanthony commented on auth/store.go:
```diff
@@ -760,13 +816,18 @@ func NewAuthStore(be backend.Backend) *authStore {
+		revision:     0,
 	}
+
+	as.initRevision(tx)
```
just `as.commitRevision(tx)`?

mitake replied:
Yes, the `initRevision()` is a needless thing. I'll remove it.

heyitsanthony commented on auth/store.go:
```diff
@@ -612,43 +664,47 @@ func (as *authStore) isOpPermitted(userName string, key, rangeEnd []byte, permTy
+		return ErrUserNotFound
+	}
+
+	if hasRootRole(u) {
```
I think it's idiomatic to return early on the error paths and have the succ
```go
if !hasRootRole(u) {
    return ErrPermissionDenied
}
return nil
```

mitake replied:
OK

heyitsanthony commented on auth/store.go:
```diff
@@ -52,6 +55,7 @@ var (
 	ErrRoleNotGranted       = errors.New("auth: role is not granted to the user")
 	ErrPermissionNotGranted = errors.New("auth: permission is not granted to the role")
 	ErrAuthNotEnabled       = errors.New("auth: authentication is not enabled")
+	ErrAuthRevisionObsolete = errors.New("auth: revision in header is obsolete")
```
`ErrAuthOldRevision`?

mitake replied:
OK

heyitsanthony commented on etcdserver/apply_auth.go:
```diff
@@ -27,8 +27,10 @@ type authApplierV3 struct {
+	mu sync.Mutex
+
+	user     string
+	revision uint64
```
replace these with `AuthInfo`?

mitake replied:
OK

heyitsanthony commented on etcdserver/etcdserverpb/raft_internal.proto:
```diff
@@ -14,6 +14,8 @@ message RequestHeader {
   // username is a username that is associated with an auth token of gRPC connection
   string username = 2;
+  // revision is a revision number of auth.authStore. It is not related to mvcc
+  uint64 revision = 3;
```
`auth_revision`?

mitake replied:
`auth_revision` seems to be good. I'll rename it in the next update.

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -87,17 +88,38 @@ func (s *EtcdServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeRe
+
+			result = s.applyV3.Apply(&pb.InternalRaftRequest{Header: hdr, Range: r})
+
+			if result.err != nil {
```
shouldn't this check for the old revision error and retry?

mitake replied:
Yes, I'll recover it.

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -109,7 +131,14 @@ func (s *EtcdServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeRe
-	result, err := s.processInternalRaftRequest(ctx, pb.InternalRaftRequest{Put: r})
+	var result *applyResult
+	var err error
+	for {
```
this loop is copied around in a lot of places; auth revision checking probably belongs in processInternalRaftRequest

mitake replied:
Yes, maybe a new function would be better for reducing indent?

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -135,16 +171,38 @@ func (s *EtcdServer) Txn(ctx context.Context, r *pb.TxnRequest) (*pb.TxnResponse
+
+			result = s.applyV3.Apply(&pb.InternalRaftRequest{Header: hdr, Txn: r})
+
+			if result.err != nil {
```
should this retry if there is an old revision error?

mitake replied:
Sorry for that, I accidentally removed the check. I'll recover it in the next update.

heyitsanthony commented:
generally looks OK, deduping the repeated retry loops would probably make this patch a lot cleaner

mitake commented:
@heyitsanthony updated for your comments, PTAL

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -562,5 +605,18 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
 	}
 }
 
+func (s *EtcdServer) processInternalRaftRequestWithRetry(ctx context.Context, r pb.InternalRaftRequest) (*applyResult, error) {
```
keep this named `processInternalRaftRequest` since the no retry case is rare / never used and rename the old `processInternalRaftRequest` to `processInternalRaftRequestOnce`?

mitake replied:
The naming seems to be better than current ones. I'll change them.

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -562,5 +605,18 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
+func (s *EtcdServer) processInternalRaftRequestWithRetry(ctx context.Context, r pb.InternalRaftRequest) (*applyResult, error) {
+	var result *applyResult
+	var err error
+	for {
```
move
```
username, err := s.usernameFromCtx(ctx)                                                       
if err != nil {
      return nil, err                                                                       
}       
r.Header.Username = username   
```
from `processInternalRaftRequest` to the start of this loop?

mitake replied:
Yes, I'll move the username lookup.

heyitsanthony commented:
lgtm after fixing last few nits. @xiang90, any thoughts on this?

xiang90 commented:
@heyitsanthony It makes sense. And the approach looks good to me.

mitake commented:
@heyitsanthony @xiang90 updated for the comments, PTAL.

heyitsanthony commented:
lgtm

xiang90 commented:
@mitake Can we fix CI?

mitake commented:
@heyitsanthony on the second thought, I think the initialization of request header's auth info shouldn't be moved to the retry loop. It is because the initialization of `r.Header` should also be moved to the loop and also to the callers of `processInternalRaftRequestOnce()`. Duplicating the code ins't good so I'd like to keep the place of the initialization. Is it ok?

heyitsanthony commented:
@mitake OK

mitake commented:
@heyitsanthony @xiang90 updated for the `datarace. authStore.commitRevisision()` missed `tx.Lock()` and `tx.Unlock()`. The race would be resolved.

xiang90 commented:
lgtm
